packages/eslint-plugin/src/util/getOperatorPrecedence.ts

```diff
@@ -0,0 +1,347 @@
+    case SyntaxKind.MetaProperty:
+      return OperatorPrecedence.Member;
+
+    case SyntaxKind.AsExpression:
+      return OperatorPrecedence.Relational;
+
+    case SyntaxKind.ThisKeyword:
+    case SyntaxKind.SuperKeyword:
+    case SyntaxKind.Identifier:
+    case SyntaxKind.PrivateIdentifier:
+    case SyntaxKind.NullKeyword:
+    case SyntaxKind.TrueKeyword:
+    case SyntaxKind.FalseKeyword:
+    case SyntaxKind.NumericLiteral:
+    case SyntaxKind.BigIntLiteral:
+    case SyntaxKind.StringLiteral:
+    case SyntaxKind.ArrayLiteralExpression:
+    case SyntaxKind.ObjectLiteralExpression:
+    case SyntaxKind.FunctionExpression:
+    case SyntaxKind.ArrowFunction:
+    case SyntaxKind.ClassExpression:
+    case SyntaxKind.RegularExpressionLiteral:
+    case SyntaxKind.NoSubstitutionTemplateLiteral:
+    case SyntaxKind.TemplateExpression:
+    case SyntaxKind.ParenthesizedExpression:
+    case SyntaxKind.OmittedExpression:
+    case SyntaxKind.JsxElement:
+    case SyntaxKind.JsxSelfClosingElement:
+    case SyntaxKind.JsxFragment:
+      return OperatorPrecedence.Primary;
+
+    default:
+      return OperatorPrecedence.Invalid;
+  }
+}
+
+export function getBinaryOperatorPrecedence(
+  kind: SyntaxKind,
+): OperatorPrecedence {
+  switch (kind) {
+    case SyntaxKind.QuestionQuestionToken:
+      return OperatorPrecedence.Coalesce;
+    case SyntaxKind.BarBarToken:
+      return OperatorPrecedence.LogicalOR;
+    case SyntaxKind.AmpersandAmpersandToken:
+      return OperatorPrecedence.LogicalAND;
+    case SyntaxKind.BarToken:
+      return OperatorPrecedence.BitwiseOR;
+    case SyntaxKind.CaretToken:
+      return OperatorPrecedence.BitwiseXOR;
+    case SyntaxKind.AmpersandToken:
+      return OperatorPrecedence.BitwiseAND;
+    case SyntaxKind.EqualsEqualsToken:
+    case SyntaxKind.ExclamationEqualsToken:
+    case SyntaxKind.EqualsEqualsEqualsToken:
+    case SyntaxKind.ExclamationEqualsEqualsToken:
+      return OperatorPrecedence.Equality;
+    case SyntaxKind.LessThanToken:
+    case SyntaxKind.GreaterThanToken:
+    case SyntaxKind.LessThanEqualsToken:
+    case SyntaxKind.GreaterThanEqualsToken:
+    case SyntaxKind.InstanceOfKeyword:
+    case SyntaxKind.InKeyword:
+    case SyntaxKind.AsKeyword:
+      return OperatorPrecedence.Relational;
+    case SyntaxKind.LessThanLessThanToken:
+    case SyntaxKind.GreaterThanGreaterThanToken:
+    case SyntaxKind.GreaterThanGreaterThanGreaterThanToken:
+      return OperatorPrecedence.Shift;
+    case SyntaxKind.PlusToken:
+    case SyntaxKind.MinusToken:
+      return OperatorPrecedence.Additive;
+    case SyntaxKind.AsteriskToken:
+    case SyntaxKind.SlashToken:
+    case SyntaxKind.PercentToken:
+      return OperatorPrecedence.Multiplicative;
+    case SyntaxKind.AsteriskAsteriskToken:
+      return OperatorPrecedence.Exponentiation;
+  }
+
+  // -1 is lower than all other precedences.  Returning it will cause binary expression
+  // parsing to stop.
+  return -1;
+}
```

Comment:
@bradzacher 
Precedence now determines whether parentheses need to be added or not.

I copied the typescript code verbatim, is that ok?
https://github.com/microsoft/TypeScript/blob/af689cc5d527789cbc1d18a7b8aae6fa46051ea7/src/compiler/utilities.ts#L3446-L3783